lib/entities.ts

```diff
@@ -1,10 +1,121 @@
 import { getCollection } from "./collection.js";
-import { HassEntities, StateChangedEvent, UnsubscribeFunc } from "./types.js";
+import {
+  Context,
+  HassEntities,
+  StateChangedEvent,
+  UnsubscribeFunc,
+} from "./types.js";
 import { Connection } from "./connection.js";
 import { Store } from "./store.js";
 import { getStates } from "./commands.js";
+import { atLeastHaVersion } from "./util.js";
 
-function processEvent(store: Store<HassEntities>, event: StateChangedEvent) {
+interface EntityState {
+  /** state */
+  s: string;
+  /** attributes */
+  a: { [key: string]: any };
+  /** context */
+  c: Context;
+  /** last_changed; if set, also applies to lu */
+  lc: string;
+  /** last_updated */
+  lu: string;
+}
+
+interface EntityDiff {
+  "+"?: Partial<EntityState>;
+  "-"?: Pick<EntityState, "a">;
+}
+
+interface StatesUpdates {
+  add?: Record<string, EntityState>;
+  removed?: string[];
+  changed: Record<string, EntityDiff>;
+}
+
+function processEvent(store: Store<HassEntities>, updates: StatesUpdates) {
+  const state = { ...store.state };
+
+  if (updates.add) {
+    for (const entityId in updates.add) {
+      const newState = updates.add[entityId];
+      state[entityId] = {
+        entity_id: entityId,
+        state: newState.s,
+        attributes: newState.a,
+        context: newState.c,
+        last_changed: newState.lc,
+        last_updated: newState.lu,
+      };
+    }
+  }
+
+  if (updates.removed) {
+    for (const entityId of updates.removed) {
+      delete state[entityId];
+    }
+  }
+
+  if (updates.changed) {
+    for (const entityId in updates.changed) {
+      let entityState = state[entityId];
+
+      if (!entityState) {
+        console.warn("Received state update for unknown entity", entityId);
+        continue;
+      }
+
+      entityState = { ...entityState };
+
+      const { "+": toAdd, "-": toRemove } = updates.changed[entityId];
+
+      const attributes =
+        toAdd?.a || toRemove?.a
+          ? { ...entityState.attributes }
+          : entityState.attributes;
+
+      if (toAdd) {
+        if (toAdd.s) {
+          entityState.state = toAdd.s;
+        }
+        if (toAdd.c) {
+          entityState.context = toAdd.c;
```

Comment:
This needs to be merged like attributes

Reply:
https://github.com/home-assistant/home-assistant-js-websocket/pull/250#issuecomment-1065577482